uvtools/dspec.py

```diff
@@ -212,63 +364,97 @@ def high_pass_fourier_filter(data, wgts, filter_size, real_delta, clean2d=False,
+def linear_filter(data, wgts, delta_data, filter_dimensions, filter_centers, filter_widths, filter_factors, cache = {}):
     '''Apply a linear delay filter to waterfall data.
+        Due to performance reasons, linear cleaning only supports separable delay-rate filters.
     Arguments:
         data: 1D or 2D (real or complex) numpy array where last dimension is frequency.
         Does not assume that weights have already been multiplied!
         wgts: real numpy array of linear multiplicative weights with the same shape as the data.
+        delta_data: float, list
+                    the width of data bins. Typically Hz: float. if 2d clean, should be 2-tuple or 2-list
+        filter_dimensions: list
+                        list of integers indicating data dimensions to filter. Must be 0, 1, or -1
         filter_centers: float, list, or 1d numpy array of delays at which to center filter windows
-                        Provide in units of (seconds)
+                        Typically in units of (seconds)
         filter_widths: float, list, or 1d numpy array of widths of each delay filtere window
                         with centers specified by filter_centers.
-                        Provide in units of (seconds)
+                        Typically in units of (seconds)
         filter_factors: float, list, or 1d numpy array of factors by which filtering should be
                         applied within each filter window specified in filter_centers and
                         filter_widths. If a float or length-1 list/ndarray is provided,
                         the same filter factor will be used in every filter window.
-        df: the width of the frequency bins in Hz: float
         cache: optional dictionary for storing pre-computed delay filter matrices.
 
     Returns:
-        2d clean residual with data filtered along the frequency direction.
-        TODO: Add functionality for linear clean in time dimension
-         -- Matrix inversion could  prohibitively expensive if kernel is not separable.
+        data: array, 2d clean residual with data filtered along the frequency direction.
+        info: dictionary with filtering parameters and a list of skipped_times and skipped_channels
+
     '''
-    if isinstance(filter_centers, np.ndarray):
-        filter_centers = list(filter_centers)
-    if isinstance(filter_widths, np.ndarray):
-        filter_widths = list(filter_widths)
-    if isinstance(filter_factors, np.ndarray):
-        filter_factors = list(filter_factors)
-    if isinstance(filter_centers, np.float):
-        filter_centers = [filter_centers]
-    if isinstance(filter_widths, np.float):
-        filter_widths = [filter_widths]
-    if isinstance(filter_factors, np.float):
-        filter_factors = [filter_factors]
+    assert isinstance(filter_dimensions, (list,tuple,int)), "filter_dimensions must be a list or tuple"
```

Comment:
the next 50 lines are quite hard to parse, please add comments for future people who might need to maintain this code. also some PEP8 non-conformity here: spaces between operator, after commas..

Reply:
I've cleaned up domentation here and added pep8 conformity.